File: packages/browser-session-mcp/src/chrome.ts
import puppeteer, { type Browser } from "puppeteer-core";

let browser: Browser | undefined;

/**
 * Connect once, reuse across calls. Reconnects transparently if the previous
 * handle disconnected (e.g. Chrome restarted).
 *
 * We don't pass `browserURL` to puppeteer.connect because Puppeteer's
 * `getWSEndpoint(browserURL)` returns whatever Chrome reports in its
 * `/json/version` response verbatim — and Chrome reports its own loopback
 * (e.g. `ws://127.0.0.1:9222/devtools/browser/<id>`). When this MCP runs
 * inside a container, that 127.0.0.1 is the container's loopback, so the
 * subsequent WebSocket connect fails with ECONNREFUSED.
 *
 * Instead we fetch /json/version ourselves, take only the path of the
 * `webSocketDebuggerUrl`, and rebuild the URL using the externally-reachable
 * host we already have. The proxy (Traefik in our setup) routes that back
 * to Chrome on the host.
 */
export async function getBrowser(browserURL: string): Promise<Browser> {
  if (browser?.connected) return browser;
  const wsEndpoint = await resolveWsEndpoint(browserURL);
  browser = await puppeteer.connect({
    browserWSEndpoint: wsEndpoint,
    defaultViewport: null,
    protocolTimeout: 180_000,
  });
  return browser;
}

export async function resolveWsEndpoint(browserURL: string): Promise<string> {
  const versionURL = new URL("/json/version", browserURL);
  const res = await fetch(versionURL.toString(), { method: "GET" });
  if (!res.ok) {
    throw new Error(
      `GET ${versionURL.toString()} returned ${res.status} ${res.statusText}`,
    );
  }
  const json = (await res.json()) as { webSocketDebuggerUrl?: string };
  const reported = json.webSocketDebuggerUrl;
  if (!reported) {
    throw new Error(
      `${versionURL.toString()} response did not include webSocketDebuggerUrl`,
    );
  }
  // Take only the path/search of what Chrome reports; keep our reachable host.
  const reportedURL = new URL(reported);
  const ws = new URL(versionURL.toString());
  ws.protocol = ws.protocol === "https:" ? "wss:" : "ws:";
  ws.pathname = reportedURL.pathname;
  ws.search = reportedURL.search;
  return ws.toString();
}

export async function closeBrowser(): Promise<void> {
  if (!browser) return;
  try {
    await browser.disconnect();
  } catch {
    // ignore
  }
  browser = undefined;
}

File: packages/browser-session-mcp/src/chrome.rs
//! Connection helpers for the persistent Chrome instance.
//!
//! We don't pass `browser_url` directly to chromiumoxide because Chrome's
//! `/json/version` returns whatever `webSocketDebuggerUrl` the browser
//! reports — typically `ws://127.0.0.1:9222/...`. When this binary runs in a
//! container or behind a proxy, that 127.0.0.1 is the wrong loopback. We fetch
//! `/json/version` ourselves, keep only the path of the reported URL, and
//! rebuild it on the externally-reachable host we already have.
use anyhow::{Context, Result, anyhow};
use chromiumoxide::Browser;
use futures::StreamExt;
use serde::Deserialize;
use tokio::task::JoinHandle;
use url::Url;

#[derive(Deserialize)]
struct VersionResponse {
    #[serde(rename = "webSocketDebuggerUrl")]
    web_socket_debugger_url: Option<String>,
}

pub async fn resolve_ws_endpoint(browser_url: &str) -> Result<String> {
    let version_url = Url::parse(browser_url)
        .with_context(|| format!("parsing BROWSER_URL {browser_url}"))?
        .join("/json/version")
        .context("building /json/version URL")?;
    let res = reqwest::get(version_url.as_str())
        .await
        .with_context(|| format!("GET {version_url}"))?;
    if !res.status().is_success() {
        return Err(anyhow!(
            "GET {version_url} returned {} {}",
            res.status().as_u16(),
            res.status().canonical_reason().unwrap_or("")
        ));
    }
    let json: VersionResponse = res
        .json()
        .await
        .with_context(|| format!("decoding /json/version body from {version_url}"))?;
    let reported = json
        .web_socket_debugger_url
        .ok_or_else(|| anyhow!("{version_url} response missing webSocketDebuggerUrl"))?;
    let reported_url = Url::parse(&reported)
        .with_context(|| format!("parsing reported webSocketDebuggerUrl {reported}"))?;

    let mut ws = version_url;
    let scheme = if ws.scheme() == "https" { "wss" } else { "ws" };
    ws.set_scheme(scheme)
        .map_err(|_| anyhow!("failed to switch scheme on {ws}"))?;
    ws.set_path(reported_url.path());
    ws.set_query(reported_url.query());
    Ok(ws.to_string())
}

/// Connect to Chrome and spawn the chromiumoxide event-loop handler. The
/// returned `JoinHandle` should be kept alive for the lifetime of `Browser`;
/// dropping it cancels the task and stops the connection.
pub async fn connect(browser_url: &str) -> Result<(Browser, JoinHandle<()>)> {
    let ws = resolve_ws_endpoint(browser_url).await?;
    let (browser, mut handler) = Browser::connect(ws.clone())
        .await
        .with_context(|| format!("connecting to {ws}"))?;
    let task = tokio::spawn(async move {
        while let Some(event) = handler.next().await {
            if let Err(err) = event {
                tracing::debug!(error = %err, "chromiumoxide handler event error");
            }
        }
    });
    Ok((browser, task))
}
